### ISSUE TITLE
Bug: [prefer-nullish-coalescing] doesn't report on functions in a ternary

### ISSUE THREAD

**Issue by OlivierZal** [typescript-eslint/typescript-eslint#10792]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=CYUwxgNghgTiAEEQBd5QFzwM7JgSwDsBzeAH3gFcDQAzQkYAbgChRJYElUAjTACgCU8ALwA%2BbLkIlyVWvSatw0OIhTww-PrCKYc%2BYkLHwACjAD2AWzxYQAHj1SylaiDoEGooTJduGLNsqcagCeupLELMxgZgQ48MggccJoWMEEYPCCmKaW1nYAbmZ4wOJGAN7M8PAA9NXwcAAOZjDIWJVo8AD8HZjBLFXcgmTkfe1aAO5QeKhgnQB0fADkUIsCXiORVbXYABZmFBDAaBBYZvUgTS3tg0LdN-C9-ZlQk9Pq80sra13PrzMfy1WQkezAAvkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KAHdJ0JpHBgAviG1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare let a: string | undefined;
declare let b: () => string | undefined;
declare let c: ((arg: string) => Promise<string | undefined>) | undefined;
declare let y: string;

const test = async (): Promise<void> => {
  // reports
  a ? a : y;
  b() || y;
  (await c?.('a')) || y;

  // should also report
  b() ? b() : y;
  (await c?.('a')) ? (await c?.('a')) : y;
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/prefer-nullish-coalescing": "warn"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Ternaries should report in the same way than equivalent `||`

### Actual Result

Doesn't report

### Additional Info

Similar to https://github.com/typescript-eslint/typescript-eslint/issues/10531 but on functions

**Comment by bradzacher** [comment]
My 2c -- it's actually correct to NOT report here because the suggested fix would be to replace 2 function calls with 1 function call. If the function has side-effects then this would change runtime behaviour.

**Comment by OlivierZal** [comment]
Hi @bradzacher, thanks for your feedback, so should `b() || y` continue to report?

**Comment by OlivierZal** [comment]
Fair enough, closing the issue.

**Comment by kirkwaiblinger** [comment]
Yeah the underlying reasoning is that it's a question of which expressions "look" side effect free, or at least idempotent in their side effects. Function calls are thought of as side-effectf-ful with no implication of idempotency[]().  _Technically_, member expressions can invoke arbitrary getters, but generally getter side effects are idempotent... The examples that come to mind for getter side effects are lazy initialization of a property and reactivity system registration (eg in Vue). Both of those have the same impact on the program if they are invoked once instead of twice.
